fins/server.go

```diff
@@ -10,51 +11,78 @@ type Server struct {
 	addr    Address
 	conn    *net.UDPConn
 	handler CommandHandler
+	dmarea []byte
+	closed bool
 }
 
-type CommandHandler func(req request) response
+type CommandHandler func(req request, dmarea []byte) response
 
-func NewServer(plcAddr Address, handler CommandHandler) (*Server, error) {
+const DM_AREA_SIZE = 32768
+
+func NewPLCSimulator(plcAddr Address) (*Server, error) {
 	s := new(Server)
 	s.addr = plcAddr
+	s.dmarea = make([]byte, DM_AREA_SIZE)
 
 	conn, err := net.ListenUDP("udp", plcAddr.udpAddress)
 	if err != nil {
 		return nil, err
 	}
 	s.conn = conn
 
-	s.handler = handler
-
-	if handler == nil {
-		s.handler = defaultHandler
-	}
-
 	go func() {
 		var buf [1024]byte
 		for {
 			rlen, remote, err := conn.ReadFromUDP(buf[:])
-			req := decodeRequest(buf[:rlen])
-			resp := s.handler(req)
+			if rlen > 0 {
+				req := decodeRequest(buf[:rlen])
+				resp := handler(req, s.dmarea)
 
-			_, err = conn.WriteToUDP(encodeResponse(resp), &net.UDPAddr{IP: remote.IP, Port: remote.Port})
+				_, err = conn.WriteToUDP(encodeResponse(resp), &net.UDPAddr{IP: remote.IP, Port: remote.Port})
+			}
 			if err != nil {
-				panic(err)
+				// do not complain when connection is closed by user
+				if !s.closed {
+					log.Fatal("Encountered error in server loop: ", err)
+				}
+				break
 			}
 		}
 	}()
 
 	return s, nil
 }
 
-func defaultHandler(r request) response {
-	fmt.Printf("Null command handler: 0x%04x\n", r.commandCode)
+// Works with only DM area, 2 byte integers
+func handler(r request, dmarea []byte) response {
```

Comment:
make handler on server type and you will not need to pass dmarea as a parameter.